core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -61,6 +62,39 @@ public static boolean hasDeletes(FileScanTask task) {
     return CloseableIterable.combine(splitTasks, tasks);
   }
 
+  /**
+   * Split files into FileScanTasks which only contain a single offset (rowGroup). For files which do not
+   * expose the offsets, use the normal split code.
+   * @param tasks Scan tasks, one per whole file to be split
+   * @param fallbackSplitSize the splitSize to use when the file does not contain explicit offsets to use
+   * @return Scan tasks, one per offset
+   */
+  public static CloseableIterable<FileScanTask> splitOnOffsets(CloseableIterable<FileScanTask> tasks,
+                                                               long fallbackSplitSize) {
+    Preconditions.checkArgument(fallbackSplitSize > 0,
+        "Invalid fallback split size (negative or 0): %s", fallbackSplitSize);
+
+    Iterable<FileScanTask> splitTasks = FluentIterable
+        .from(tasks)
+        .transformAndConcat(input -> {
+          DataFile file = input.file();
+          if (file.format().hasOffsets()) {
+            if (file.splitOffsets() != null) {
+              // Split on offsets, size 0 means 1 task per offset
+              return input.split(0);
+            } else {
+              // File too small to have offsets, take the entire file as a task
```

Comment:
Seems like we could just use the original split method here, just in case there are no split locations (which are optional) but the file isn't small. I don't think we need to make this assumption.

Reply:
This is again where I wanted to differentiate between globally splittable and only splittable on offsets. Originally I just used the base method but then I have to pass through the file size to get a single split which seemed odd. We have to pass through the size because we go through the non-offset split path which can possibly give us multiple splits (some of which would be empty.)

Reply:
Just want to combine the original split code here to have a general view, basically we get:

```java
if (file.format().hasOffsets()) {
    if (file.splitOffsets() != null) {
        // split to 1 task per offset
    } else {
       // file too small, use 1 task for entire file
    }
} else {
    if (file.format().isSplittable()) {
        if (file.splitOffsets() != null) {
            // offset aware split
        } else {
            // fixed size split
        }
    } else {
        // use 1 task for entire file
    }
}
```

So the case of `hasOffset && splitOffsets == null` case is basically a short-circuit even if the file is splittable, there is still some value for that.

The question I have is that would it be worth directly adding the logic to the original splitFiles, instead of having 2 different methods? I imagine this would also help reducing the number of file scan tasks generated during scan planning.

Reply:
I think it can be a little simpler
```java
if (file.format().hasOffsets()) {
    if (file.splitOffsets() != null) {
        // split to 1 task per offset
    } else {
       // file too small, use 1 task for entire file
    }
} else {
    if (file.format().isSplittable()) {
       // fixed size split (this cannot be an offset split because that would have been in the first branch)
      }
    } else {
        // use 1 task for entire file
    }
}
```

But yeah I think we can have this in the original logic, I was just afraid of changing an existing behavior mostly.

Reply:
I think that logic is incorrect. If a file doesn't have offsets, then we can't assume that it is small. It could just be that a writer didn't produce the offsets. The Parquet import code may or may not produce offsets, for example. We can't assume something about the file just because metadata is missing.

Reply:
I decided to forgo this route and try something we were discussing a bit earlier. Instead of trying to split offset files exactly on split-size boundaries, I'm going to split all files with offsets on their offsets and then recombine them when they are put in a CombinedScanTask.